lib/Core/Search/Common/PageTextExtractor.php

```diff
@@ -0,0 +1,269 @@
+use Symfony\Component\Routing\RouterInterface;
+use function count;
+use function explode;
+use function in_array;
+use function libxml_use_internal_errors;
+use function mb_strlen;
+use function mb_strpos;
+use function mb_substr;
+use function sprintf;
+use function trim;
+use const XML_ELEMENT_NODE;
+use const XML_HTML_DOCUMENT_NODE;
+use const XML_TEXT_NODE;
+
+class PageTextExtractor
+{
+    /** @var array<int, array<string, array<string, array<int, string>|string>>> */
+    private array $cache = [];
+
+    private LoggerInterface $logger;
+
+    /**
+     * @param array<string, int> $siteRoots
+     * @param array<string, array<string, string>> $languageAccessibility
+     * @param array<array<int, string>> $pageTextConfig
+     */
+    public function __construct(
+        private readonly ContentHandler $contentHandler,
+        private readonly LocationHandler $locationHandler,
+        private readonly RouterInterface $router,
+        private readonly array $siteRoots,
+        private readonly array $languageAccessibility,
+        private readonly string $pageIndexingHost,
+        private readonly array $pageTextConfig,
+    ) {
+        $this->logger = new NullLogger();
+    }
+
+    public function setLogger(LoggerInterface $logger): void
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param int $contentId
+     * @param string $languageCode
+     *
+     * @return array<string, array<int, string>|string>
+     */
+    public function extractPageText(int $contentId, string $languageCode): array
+    {
+        if (isset($this->cache[$contentId][$languageCode])) {
+            return $this->cache[$contentId][$languageCode];
+        }
+
+        if (count($this->cache) > 10) {
+            $this->cache = [];
+        }
+
+        try {
+            $html = $this->fetchPageSource($contentId, $languageCode);
+        } catch (IndexPageUnavailableException|RuntimeException $e) {
+            $this->logger->error($e->getMessage());
+
+            return [];
+        }
+
+        $textArray = $this->extractTextArray($html);
+
+        $this->cache[$contentId][$languageCode] = $textArray;
+
+        return $textArray;
+    }
+
+    /**
+     * @param string $languageCode
+     * @param int $contentId
+     *
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\NotFoundException
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\UnauthorizedException
+     *
+     * @return string
+     */
+    private function generateUrl(string $languageCode, int $contentId): string
+    {
+        $contentInfo = $this->contentHandler->loadContentInfo($contentId);
+        $siteAccess = $this->resolveSiteAccess($contentInfo, $languageCode);
+
+        $relativePath = $this->router->generate(
+            'ibexa.url.alias',
+            [
+                'locationId' => (int) $contentInfo->mainLocationId,
+                'siteaccess' => $siteAccess,
+            ],
+            UrlGeneratorInterface::RELATIVE_PATH,
```

Comment:
Let's get rid of the `host` configuration and generate an absolute URL here, that will be sufficient for most use cases. When it's tested to work OK, we will expand it by extracting URL generation to a service, so a different implementation can be injected for specific cases.

Reply:
resolved in https://github.com/netgen/ibexa-search-extra/pull/10/commits/963bbc52b6e9c1610ce956dbad0e797e86b34f5b

Reply:
resolved in f9ce3aa3f8802976be7eac51d95aad9cc304f39a